From 4c05bf68b7f2a22ade17fdc4037b4e65671d2795 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Wed, 6 May 2020 17:44:56 +0800
Subject: [PATCH] DRV_DPM_SW-236: Add locking for all QOS related APIs

- All exported APIs from datapath_qos.c are handled in
  gswip31/datapath_ppv4_api.c under a switch statement
  of qos_platform_set().
- So, use new qos_lock around the switch statement itself to cover
  all exported APIs.

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/gswip31/datapath_misc.h     | 1 +
 drivers/net/datapath/dpm/gswip31/datapath_ppv4.c     | 1 +
 drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c | 4 ++++
 3 files changed, 6 insertions(+)

diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.h b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
index b068079779a3..3a68baae809d 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.h
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
@@ -115,6 +115,7 @@ struct hal_priv {
 			 */
 	struct ppv4_port_map_table ppv4_port_map;/* PPV4 Ports table map */
 	DP_TX_CB_LOCK_T cb_lock;
+	DP_LOCK_T qos_lock;
 };
 
 static inline bool is_qid_valid(int q_id)
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c
index 6ef25c3501c4..fdd4f1fd53ec 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c
@@ -550,6 +550,7 @@ int init_ppv4_qos(int inst, int flag)
 		 priv->cqm_drop_p,
 		 priv->ppv4_drop_p);
 #endif /* end of CONFIG_INTEL_DATAPATH_QOS_HAL */
+	DP_LOCK_INIT(&priv->qos_lock);
 	DP_DEBUG(DP_DBG_FLAG_DBG, "%s done\n", __func__);
 	res = DP_SUCCESS;
 
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index 15b7d2ecb029..f7c905bb9dca 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -51,6 +51,8 @@ int qos_platform_set(int cmd_id, void *node, int flag)
 		return DP_FAILURE;
 	}
 
+	DP_LIB_LOCK(&priv->qos_lock);
+
 	switch (cmd_id) {
 	case NODE_LINK_ADD:
 		res = dp_node_link_add_31((struct dp_node_link *)node, flag);
@@ -145,6 +147,8 @@ int qos_platform_set(int cmd_id, void *node, int flag)
 		pr_err("no support yet cmd_id %d\n", cmd_id);
 		break;
 	}
+
+	DP_LIB_UNLOCK(&priv->qos_lock);
 	return res;
 }
 
